### PR TITLE
fix torch.py when saving checkpoints using pytorch2.4

### PR DESCRIPTION
**Pull request by 1195343015**
issue #1003 
this bug is due to the calling of function _add_non_coordinator_iobytes_request() in create_local_plan()
https://github.com/NVIDIA/Megatron-LM/blob/6bf8448ba065a0a37b2b874f49fd65ca9547b5c0/megatron/core/dist_checkpointing/strategies/torch.py#L445-L447
https://github.com/NVIDIA/Megatron-LM/blob/6bf8448ba065a0a37b2b874f49fd65ca9547b5c0/megatron/core/dist_checkpointing/strategies/torch.py#L469-L474
function create_default_local_save_plan() is different between Pytorch 2.3 and 2.4 
PyTorch 2.3
https://github.com/pytorch/pytorch/blob/63d5e9221bedd1546b7d364b5ce4171547db12a9/torch/distributed/checkpoint/default_planner.py#L259
```python
def create_default_local_save_plan(
    state_dict: Dict[str, Any], is_coordinator: bool
) -> SavePlan:
    """
    Create the ``SavePlan`` used by DefaultSavePlanner.

    On non-coordinator ranks, this function ignores tensors and non-tensor objects,
    only producing writes for ShardedTensor objects.

    On the coordinator rank, produce writes for all values.
    """
    requests = []
    for fqn, obj in state_dict.items():
        # Since DTensor supports submesh, adding extra check to ensure _create_write_items()
        # gets called only when the current rank is part of the mesh for the corresponding DTensor.
        if isinstance(obj, DTensor):
            if obj.device_mesh.get_coordinate() is not None:
                requests += _create_write_items(fqn, obj)
        elif isinstance(obj, (torch.Tensor)) or is_coordinator:
            requests += _create_write_items(fqn, obj)

    return SavePlan(requests)

``` 
Pytorch 2.4
https://github.com/pytorch/pytorch/blob/f6fb80b0f906ebc89e4bfcfe4e322585cbbcecb8/torch/distributed/checkpoint/default_planner.py#L344
```python
def create_default_local_save_plan(
    state_dict: Dict[str, Any], is_coordinator: bool
) -> SavePlan:
    """
    Create the ``SavePlan`` used by DefaultSavePlanner.

    On non-coordinator ranks, this function ignores tensors and non-tensor objects,
    only producing writes for ShardedTensor objects.

    On the coordinator rank, produce writes for all values.
    """
    requests = []
    for fqn, obj in state_dict.items():
        # Since DTensor supports submesh, adding extra check to ensure _create_write_items()
        # gets called only when the current rank is part of the mesh for the corresponding DTensor.
        if isinstance(obj, DTensor):
            if obj.device_mesh.get_coordinate() is not None:
                requests += _create_write_items(fqn, obj)
        else:
            # For the plain tensor and non-tensor values, add the request for all
            # the ranks. Coordinator will decides whether to deduplicate the
            # values based on the keys.
            requests += _create_write_items(fqn, obj)

    return SavePlan(requests)
``` 

This bug happens when using pytorch2.4. Because in pytorch 2.4, `is_coordinator` condition check is removed. `non_coordinator` requests haved been added in "plan" by PyTorch. So it's no need to call `_add_non_coordinator_iobytes_request` in Megatron core.

### CODE REVIEW

**Comment by akoumpa** [comment]
CC @mikolajblaz

**Comment by mikolajblaz** [comment]
Thanks @1195343015 for providing pointers to this silent change on the PyT side!
Let me incorporate your fix

**Comment by MangoFF** [comment]
I follow @1195343015  change the code, it seem that the loss while continue to train is different from before , I check the other is the same.

**Comment by MangoFF** [comment]
Is anyone check this is right?

**Comment by mikolajblaz** [comment]
Merged and available in MCore v0.9 (in a slightly different formulation: https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/core/dist_checkpointing/strategies/torch.py#L460)
